Currently Abins identifies if the mass is a standard isotopic mixture by comparing against `Atom(symbol=symbol).mass` with a threshold of 0.01.

In the case of zinc, the standard mass has been revised by a difference of more than 0.01, so CASTEP writes mass values that are not detected by Abins as being the standard.

A more robust workflow could look at which is _nearest_ between the standard mass and another known isotope.

I have found a quirk the `Atom` API though:

```
In [44]: Atom(symbol="Zn").mass
Out[44]: 65.409

In [45]: Atom(symbol="Zn", a_number=82).mass
Out[45]: 81.95484

In [46]: Atom(symbol="Zn", a_number=83).mass
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[46], line 1
----> 1 Atom(symbol="Zn", a_number=83).mass

RuntimeError: Failed to find an atom with symbol=Zn and a=83

In [47]: Atom(symbol="Zn", a_number=20).mass
Out[47]: 53.99295
```

Mass values that are too high raise an error, but masses that are too _low_ do not!

--- a/Framework/PythonInterface/plugins/algorithms/Abins.py
+++ b/Framework/PythonInterface/plugins/algorithms/Abins.py
@@ -7,6 +7,8 @@
 
 from typing import Dict
 
+import numpy as np
+
 from mantid.api import AlgorithmFactory, FileAction, FileProperty, PythonAlgorithm, Progress
 from mantid.api import WorkspaceFactory, AnalysisDataService
 
@@ -14,7 +16,7 @@ from mantid.api import WorkspaceFactory, AnalysisDataService
 from mantid.simpleapi import ConvertUnits, GroupWorkspaces, Load
 from mantid.kernel import Direction, StringListValidator
 import abins
-from abins.abinsalgorithm import AbinsAlgorithm
+from abins.abinsalgorithm import AbinsAlgorithm, AtomInfo
 
 
 # noinspection PyPep8Naming,PyMethodMayBeStatic
@@ -184,14 +186,13 @@ class Abins(AbinsAlgorithm, PythonAlgorithm):
         self.setProperty("OutputWorkspace", self._out_ws_name)
         prog_reporter.report("Group workspace with all required  dynamical structure factors has been constructed.")
 
-    def _fill_s_workspace(self, s_points=None, workspace=None, protons_number=None, nucleons_number=None):
+    def _fill_s_workspace(self, *, s_points: np.ndarray, workspace: str, species: AtomInfo | None = None):
         """
         Puts S into workspace(s).
 
         :param s_points: dynamical factor for the given atom
         :param workspace:  workspace to be filled with S
-        :param protons_number: number of protons in the given type fo atom
-        :param nucleons_number: number of nucleons in the given type of atom
+        :param species: atom/isotope identity and data
         """
 
         from abins.constants import FUNDAMENTALS, ONE_DIMENSIONAL_INSTRUMENTS, ONE_DIMENSIONAL_SPECTRUM
@@ -201,15 +202,11 @@ class Abins(AbinsAlgorithm, PythonAlgorithm):
 
         # only FUNDAMENTALS [data is 2d with one row]
         if s_points.shape[0] == FUNDAMENTALS:
-            self._fill_s_1d_workspace(
-                s_points=s_points[0], workspace=workspace, protons_number=protons_number, nucleons_number=nucleons_number
-            )
+            self._fill_s_1d_workspace(s_points=s_points[0], workspace=workspace, species=species)
 
         # total workspaces [data is 1d vector]
         elif len(s_points.shape) == ONE_DIMENSIONAL_SPECTRUM:
-            self._fill_s_1d_workspace(
-                s_points=s_points, workspace=workspace, protons_number=protons_number, nucleons_number=nucleons_number
-            )
+            self._fill_s_1d_workspace(s_points=s_points, workspace=workspace, species=species)
 
         # quantum order events (fundamentals  or  overtones + combinations for the given order)
         # [data is 2d table of S with a row for each quantum order]
@@ -221,26 +218,23 @@ class Abins(AbinsAlgorithm, PythonAlgorithm):
                 wrk_name = f"{workspace}_quantum_event_{n + 1}"
                 partial_wrk_names.append(wrk_name)
 
-                self._fill_s_1d_workspace(
-                    s_points=s_points[n], workspace=wrk_name, protons_number=protons_number, nucleons_number=nucleons_number
-                )
+                self._fill_s_1d_workspace(s_points=s_points[n], workspace=wrk_name, species=species)
 
             GroupWorkspaces(InputWorkspaces=partial_wrk_names, OutputWorkspace=workspace)
 
-    def _fill_s_1d_workspace(self, s_points=None, workspace=None, protons_number=None, nucleons_number=None):
+    def _fill_s_1d_workspace(self, *, s_points: np.ndarray, workspace: str, species: AtomInfo | None = None):
         """
         Puts 1D S into workspace.
-        :param protons_number: number of protons in the given type of atom
-        :param nucleons_number: number of nucleons in the given type of atom
         :param s_points: dynamical factor for the given atom
         :param workspace: workspace to be filled with S
+        :param species: atom/isotope identity and data
         """
-        if protons_number is not None:
+        if species is not None:
             s_points = (
                 s_points
                 * self._scale
                 * self.get_cross_section(
-                    scattering=self._scale_by_cross_section, protons_number=protons_number, nucleons_number=nucleons_number
+                    scattering=self._scale_by_cross_section, protons_number=species.z_number, nucleons_number=species.nucleons_number
                 )
             )
         dim = 1

--- a/scripts/abins/abins2.py
+++ b/scripts/abins/abins2.py
@@ -7,13 +7,15 @@
 
 from typing import Dict, Optional
 
+import numpy as np
+
 from mantid.api import AlgorithmFactory, PythonAlgorithm, Progress
 from mantid.api import WorkspaceFactory, AnalysisDataService
 
 # noinspection PyProtectedMember
 from mantid.simpleapi import ConvertUnits, GroupWorkspaces
 import abins
-from abins.abinsalgorithm import AbinsAlgorithm
+from abins.abinsalgorithm import AbinsAlgorithm, AtomInfo
 from abins.logging import get_logger, Logger
 
 
@@ -170,14 +172,13 @@ class Abins(AbinsAlgorithm, PythonAlgorithm):
         self.setProperty("OutputWorkspace", self._out_ws_name)
         prog_reporter.report("Group workspace with all required  dynamical structure factors has been constructed.")
 
-    def _fill_s_workspace(self, s_points=None, workspace=None, protons_number=None, nucleons_number=None):
+    def _fill_s_workspace(self, *, s_points: np.ndarray, workspace: str, species: AtomInfo | None = None) -> None:
         """
         Puts S into workspace(s).
 
         :param s_points: dynamical factor for the given atom
-        :param workspace:  workspace to be filled with S
-        :param protons_number: number of protons in the given type fo atom
-        :param nucleons_number: number of nucleons in the given type of atom
+        :param workspace: workspace to be filled with S
+        :param species: Element/isotope data
         """
 
         from abins.constants import FUNDAMENTALS, ONE_DIMENSIONAL_INSTRUMENTS, ONE_DIMENSIONAL_SPECTRUM
@@ -187,15 +188,11 @@ class Abins(AbinsAlgorithm, PythonAlgorithm):
 
         # only FUNDAMENTALS [data is 2d with one row]
         if s_points.shape[0] == FUNDAMENTALS:
-            self._fill_s_1d_workspace(
-                s_points=s_points[0], workspace=workspace, protons_number=protons_number, nucleons_number=nucleons_number
-            )
+            self._fill_s_1d_workspace(s_points=s_points[0], workspace=workspace, species=species)
 
         # total workspaces [data is 1d vector]
         elif len(s_points.shape) == ONE_DIMENSIONAL_SPECTRUM:
-            self._fill_s_1d_workspace(
-                s_points=s_points, workspace=workspace, protons_number=protons_number, nucleons_number=nucleons_number
-            )
+            self._fill_s_1d_workspace(s_points=s_points, workspace=workspace, species=species)
 
         # quantum order events (fundamentals  or  overtones + combinations for the given order)
         # [data is 2d table of S with a row for each quantum order]
@@ -207,13 +204,11 @@ class Abins(AbinsAlgorithm, PythonAlgorithm):
                 wrk_name = f"{workspace}_quantum_event_{n + 1}"
                 partial_wrk_names.append(wrk_name)
 
-                self._fill_s_1d_workspace(
-                    s_points=s_points[n], workspace=wrk_name, protons_number=protons_number, nucleons_number=nucleons_number
-                )
+                self._fill_s_1d_workspace(s_points=s_points[n], workspace=wrk_name, species=species)
 
             GroupWorkspaces(InputWorkspaces=partial_wrk_names, OutputWorkspace=workspace)
 
-    def _fill_s_1d_workspace(self, s_points=None, workspace=None, protons_number=None, nucleons_number=None):
+    def _fill_s_1d_workspace(self, s_points=None, workspace=None, species: AtomInfo | None = None):
         """
         Puts 1D S into workspace.
         :param protons_number: number of protons in the given type of atom
@@ -221,9 +216,9 @@ class Abins(AbinsAlgorithm, PythonAlgorithm):
         :param s_points: dynamical factor for the given atom
         :param workspace: workspace to be filled with S
         """
-        if protons_number is not None:
+        if species is not None:
             s_points = s_points * self.get_cross_section(
-                scattering=self._scale_by_cross_section, protons_number=protons_number, nucleons_number=nucleons_number
+                scattering=self._scale_by_cross_section, protons_number=species.z_number, nucleons_number=species.nucleons_number
             )
         dim = 1
         length = s_points.size

--- a/scripts/abins/abinsalgorithm.py
+++ b/scripts/abins/abinsalgorithm.py
@@ -480,10 +480,9 @@ class AbinsAlgorithm:
 
         ws_name = self._out_ws_name + "_" + label
         self._fill_s_workspace(
+            species=species,
             s_points=s_points,
             workspace=ws_name,
-            protons_number=species.z_number,
-            nucleons_number=species.nucleons_number,
         )
         return ws_name
 
@@ -639,7 +638,7 @@ class AbinsAlgorithm:
                         s_atoms[:, i] += mtd[partial_ws].dataY(i)
 
             # create workspace with S
-            self._fill_s_workspace(s_atoms, total_workspace)
+            self._fill_s_workspace(s_points=s_atoms, workspace=total_workspace)
 
         # # Otherwise just repackage the workspace we have as the total
         else:

--- a/scripts/abins/abinsalgorithm.py
+++ b/scripts/abins/abinsalgorithm.py
@@ -7,6 +7,8 @@
 
 # Supporting functions for the Abins Algorithm that don't belong in
 # another part of AbinsModules.
+import dataclasses
+from functools import cached_property
 from math import isnan
 import os
 from pathlib import Path
@@ -29,6 +31,38 @@ from abins.constants import AB_INITIO_FILE_EXTENSIONS, ALL_INSTRUMENTS, ATOM_PRE
 from abins.input.jsonloader import abins_supported_json_formats, JSONLoader
 from abins.instruments import get_instrument, Instrument
 import abins.parameters
+
+
+@dataclasses.dataclass
+class AtomInfo:
+    symbol: str
+    mass: float
+
+    @cached_property
+    def name(self):
+        if self.nucleons_number:
+            return f"{self.nucleons_number}{self.symbol}"
+        return self.symbol
+
+    @property
+    def z_number(self):
+        return self._mantid_atom.z_number
+
+    @property
+    def nucleons_number(self):
+        return self._mantid_atom.a_number
+
+    @cached_property
+    def neutron_data(self):
+        return self._mantid_atom.neutron()
+
+    @cached_property
+    def _mantid_atom(self):
+        nearest_int = int(round(self.mass))
+        nearest_isotope = Atom(symbol=self.symbol, a_number=nearest_int)
+        standard_mix = Atom(symbol=self.symbol)
+
+        return min((nearest_isotope, standard_mix), key=(lambda atom: abs(atom.mass - self.mass)))
 
 
 class AbinsAlgorithm:
@@ -398,7 +432,7 @@ class AbinsAlgorithm:
 
         :returns: workspaces for list of atoms types, S for the particular type of atom
         """
-        from abins.constants import FLOAT_TYPE, MASS_EPS, ONLY_ONE_MASS
+        from abins.constants import FLOAT_TYPE
 
         # Create appropriately-shaped arrays to be used in-place by _atom_type_s - avoid repeated slow instantiation
         shape = [max_quantum_order]
@@ -413,7 +447,6 @@ class AbinsAlgorithm:
 
         if atoms_symbols is not None:
             for symbol in atoms_symbols:
-                sub = len(masses[symbol]) > ONLY_ONE_MASS or abs(Atom(symbol=symbol).mass - masses[symbol][0]) > MASS_EPS
                 for m in masses[symbol]:
                     result.extend(
                         self._atom_type_s(
@@ -424,7 +457,6 @@ class AbinsAlgorithm:
                             element_symbol=symbol,
                             temp_s_atom_data=temp_s_atom_data,
                             s_atom_data=s_atom_data,
-                            substitution=sub,
                         )
                     )
         if atom_numbers is not None:
@@ -475,8 +507,7 @@ class AbinsAlgorithm:
         atom_workspaces = []
         s_atom_data.fill(0.0)
         output_atom_label = "%s_%d" % (ATOM_PREFIX, atom_number)
-        symbol = atoms_data[atom_number - 1]["symbol"]
-        z_number = Atom(symbol=symbol).z_number
+        atom = Atom(atoms_data[atom_number - 1]["symbol"])
 
         for i, order in enumerate(range(FUNDAMENTALS, self._max_event_order + 1)):
             s_atom_data[i] = s_data[atom_number - 1]["order_%s" % order]
@@ -486,10 +517,12 @@ class AbinsAlgorithm:
         atom_workspaces = []
         atom_workspaces.append(
             self._create_workspace(
-                atom_name=output_atom_label, s_points=np.copy(total_s_atom_data), optional_name="_total", protons_number=z_number
+                atom_name=output_atom_label, s_points=np.copy(total_s_atom_data), optional_name="_total", protons_number=atom.z_number
             )
         )
-        atom_workspaces.append(self._create_workspace(atom_name=output_atom_label, s_points=np.copy(s_atom_data), protons_number=z_number))
+        atom_workspaces.append(
+            self._create_workspace(atom_name=output_atom_label, s_points=np.copy(s_atom_data), protons_number=atom.z_number)
+        )
         return atom_workspaces
 
     def _atom_type_s(
@@ -501,7 +534,6 @@ class AbinsAlgorithm:
         element_symbol=None,
         temp_s_atom_data=None,
         s_atom_data=None,
-        substitution=None,
     ):
         """
         Helper function for calculating S for the given type of atom
@@ -516,16 +548,16 @@ class AbinsAlgorithm:
             information but is used in-place to save on time instantiating large arrays.
         :param s_atom_data: helper array to accumulate S (outer loop over atoms); does not transport
             information but is used in-place to save on time instantiating large arrays.
-        :param substitution: True if isotope substitution and False otherwise
         """
         from abins.constants import MASS_EPS
 
         atom_workspaces = []
         s_atom_data.fill(0.0)
 
-        element = Atom(symbol=element_symbol)
+        species = AtomInfo(symbol=element_symbol, mass=mass)
 
         for atom_index in range(num_atoms):
+            # TODO this mass_eps could be smaller, we are checking against known masses from data
             if atoms_data[atom_index]["symbol"] == element_symbol and abs(atoms_data[atom_index]["mass"] - mass) < MASS_EPS:
                 temp_s_atom_data.fill(0.0)
 
@@ -538,35 +570,23 @@ class AbinsAlgorithm:
 
         total_s_atom_data = np.sum(s_atom_data, axis=0)
 
-        nucleons_number = int(round(mass))
-
-        if substitution:
-            atom_workspaces.append(
-                self._create_workspace(
-                    atom_name=str(nucleons_number) + element_symbol,
-                    s_points=np.copy(total_s_atom_data),
-                    optional_name="_total",
-                    protons_number=element.z_number,
-                    nucleons_number=nucleons_number,
-                )
+        atom_workspaces.append(
+            self._create_workspace(
+                atom_name=species.name,
+                s_points=np.copy(total_s_atom_data),
+                optional_name="_total",
+                protons_number=species.z_number,
+                nucleons_number=species.nucleons_number,
             )
-            atom_workspaces.append(
-                self._create_workspace(
-                    atom_name=str(nucleons_number) + element_symbol,
-                    s_points=np.copy(s_atom_data),
-                    protons_number=element.z_number,
-                    nucleons_number=nucleons_number,
-                )
+        )
+        atom_workspaces.append(
+            self._create_workspace(
+                atom_name=species.name,
+                s_points=np.copy(s_atom_data),
+                protons_number=species.z_number,
+                nucleons_number=species.nucleons_number,
             )
-        else:
-            atom_workspaces.append(
-                self._create_workspace(
-                    atom_name=element_symbol, s_points=np.copy(total_s_atom_data), optional_name="_total", protons_number=element.z_number
-                )
-            )
-            atom_workspaces.append(
-                self._create_workspace(atom_name=element_symbol, s_points=np.copy(s_atom_data), protons_number=element.z_number)
-            )
+        )
 
         return atom_workspaces
 

--- a/scripts/abins/abinsalgorithm.py
+++ b/scripts/abins/abinsalgorithm.py
@@ -464,22 +464,27 @@ class AbinsAlgorithm:
                 result.extend(self._atom_number_s(atom_number=atom_number, s_data=s_data, s_atom_data=s_atom_data, atoms_data=atoms_data))
         return result
 
-    def _create_workspace(self, atom_name=None, s_points=None, optional_name="", protons_number=None, nucleons_number=None):
+    def _create_workspace(self, *, species: AtomInfo, s_points: np.ndarray, label: str | None = None):
         """
         Creates workspace for the given frequencies and s_points with S data. After workspace is created it is rebined,
         scaled by cross-section factor and optionally multiplied by the user defined scaling factor.
 
 
-        :param atom_name: symbol of atom for which workspace should be created
+        :param species: Object identifying isotope (or mixture)
         :param s_points: S(Q, omega)
-        :param optional_name: optional part of workspace name
+        :param label: species-specific part of workspace name. If None, take from species object.
         :returns: workspace for the given frequency and S data
-        :param protons_number: number of protons in the given type fo atom
-        :param nucleons_number: number of nucleons in the given type of atom
         """
 
-        ws_name = self._out_ws_name + "_" + atom_name + optional_name
-        self._fill_s_workspace(s_points=s_points, workspace=ws_name, protons_number=protons_number, nucleons_number=nucleons_number)
+        label = species.name if label is None else label
+
+        ws_name = self._out_ws_name + "_" + label
+        self._fill_s_workspace(
+            s_points=s_points,
+            workspace=ws_name,
+            protons_number=species.z_number,
+            nucleons_number=species.nucleons_number,
+        )
         return ws_name
 
     def _atom_number_s(self, *, atom_number, s_data, s_atom_data, atoms_data):
@@ -507,7 +512,8 @@ class AbinsAlgorithm:
         atom_workspaces = []
         s_atom_data.fill(0.0)
         output_atom_label = "%s_%d" % (ATOM_PREFIX, atom_number)
-        atom = Atom(atoms_data[atom_number - 1]["symbol"])
+        atom_data = atoms_data[atom_number - 1]
+        species = AtomInfo(symbol=atom_data["symbol"], mass=atom_data["mass"])
 
         for i, order in enumerate(range(FUNDAMENTALS, self._max_event_order + 1)):
             s_atom_data[i] = s_data[atom_number - 1]["order_%s" % order]
@@ -517,12 +523,12 @@ class AbinsAlgorithm:
         atom_workspaces = []
         atom_workspaces.append(
             self._create_workspace(
-                atom_name=output_atom_label, s_points=np.copy(total_s_atom_data), optional_name="_total", protons_number=atom.z_number
+                species=species,
+                s_points=np.copy(total_s_atom_data),
+                label=output_atom_label + "_total",
             )
         )
-        atom_workspaces.append(
-            self._create_workspace(atom_name=output_atom_label, s_points=np.copy(s_atom_data), protons_number=atom.z_number)
-        )
+        atom_workspaces.append(self._create_workspace(species=species, s_points=np.copy(s_atom_data), label=output_atom_label))
         return atom_workspaces
 
     def _atom_type_s(
@@ -572,19 +578,15 @@ class AbinsAlgorithm:
 
         atom_workspaces.append(
             self._create_workspace(
-                atom_name=species.name,
+                species=species,
                 s_points=np.copy(total_s_atom_data),
-                optional_name="_total",
-                protons_number=species.z_number,
-                nucleons_number=species.nucleons_number,
+                label=f"{species.name}_total",
             )
         )
         atom_workspaces.append(
             self._create_workspace(
-                atom_name=species.name,
+                species=species,
                 s_points=np.copy(s_atom_data),
-                protons_number=species.z_number,
-                nucleons_number=species.nucleons_number,
             )
         )
 

